spec/mempool/gossip/dog.md

```diff
@@ -0,0 +1,606 @@
+    > not natively support specifying time constraints without relying on workarounds such as using
+    > counters as clocks. As a result, this action is modelled as always being enabled. In a
+    > real-world implementation, redundancy adjustments would be naturally triggered by a timer set
+    > to the duration specified by `adjustInterval`.
+    
+### Adding transactions to the mempool
+
+`tryAddTx` defines how a node adds a transaction to its mempool. The following code is the same as
+in Flood; the difference is in the two functions that process the transaction.
+```bluespec "actions" +=
+action tryAddTx(node, _incomingMsgs, optionalSender, tx) = 
+    if (not(hash(tx).in(node.Cache())))
+        node.tryAddFirstTimeTx(_incomingMsgs, optionalSender, tx)
+    else
+        node.processDuplicateTx(_incomingMsgs, optionalSender, tx)
+```
+
+* **Adding first-time transactions**
+
+    `tryAddFirstTimeTx` attempts to add a received first-time transaction `tx` to the mempool by
+    performing the same updates as in Flood (it adds `tx` to `cache`; if `tx` is valid, it appends `tx`
+    to the mempool and updates `tx`'s senders). Additionally, it increases the `node`'s
+    `rc.firstTimeTxs` counter.
+    ```bluespec "actions" +=
+    action tryAddFirstTimeTx(node, _incomingMsgs, optionalSender, tx) = 
+        all {
+            node.Flood::tryAddFirstTimeTx(_incomingMsgs, optionalSender, tx),
+            rc' = rc.update(node, increaseFirstTimeTxs),
+            dr' = dr,
+        }
+    ```
+
+* **Handling duplicate transactions**
+
+    `processDuplicateTx` processes a received duplicate transaction `tx` by updating the list of senders
+    if `tx` is in the mempool, and the list of incoming messages, the same as in Flood. Additionally,
+    1. it increases `duplicateTxs` and 
+    2. replies a `HaveTx` message if the RC mechanism is not blocking it (and there's a sender). 
+
+    ```bluespec "actions" +=
+    action processDuplicateTx(node, _incomingMsgs, optionalSender, tx) =
+        val _rc = node.RC().increaseDuplicateTxs()
+        val updatedVars = node.replyHaveTx(_incomingMsgs, _rc, optionalSender, tx)
+        val _incomingMsgs1 = updatedVars._1
+        val _rc1 = updatedVars._2
+        all {
+            node.Flood::processDuplicateTx(_incomingMsgs1, optionalSender, tx),
+            rc' = rc.put(node, _rc1),
+            dr' = dr,
+        }
+    ```
+    where `replyHaveTx` will send a `HaveTx` message if `tx` comes from a peer and `HaveTx` messages are
+    not blocked:
+    ```bluespec "actions" +=
+    pure def replyHaveTx(node, _incomingMsgs, _rc, optionalSender, tx) =
+        if (optionalSender.isSome() and not(_rc.isHaveTxBlocked))
+            val targets = optionalSender.optionToSet()
+            (node.multiSend(_incomingMsgs, targets, HaveTxMsg(hash(tx))), _rc.blockHaveTx())
+        else (_incomingMsgs, _rc)
+    ```
+    Immediately after sending the `HaveTx` message the controller will block sending a new one.
```

Comment:
This is an information that is only introduced here. I realized that before by investigating the Quint code.

This is not a detail and should be clear from the beginning: at each cycle of the redundancy control mechanism, the node can send at most one `HaveTx` message.

Reply:
Thanks for pointing out. I made it more clear in a couple of places in the Redundancy Control section.